src/mongo_odm/test/query_builder.cpp

```diff
@@ -162,10 +602,456 @@ TEST_CASE("Query builder works with non-ODM class") {
+                              MONGO_ODM_KEY(Bar::pts).pull(MONGO_ODM_KEY(Point::x) > 10));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 1);
+
+        bar = Bar::find_one(MONGO_ODM_KEY(Bar::w) == 888);
+        REQUIRE(bar);
+        REQUIRE((bar->pts == std::vector<Point>{{9, 10}}));
+    }
+
+    SECTION("Test $pullAll array update operator", "[mongo_odm::Nvp::pull]") {
+        Bar(1010, 10, 999, true, "pullAll", {1, 0}, {4, 5, 6}, {{9, 10}, {11, 12}},
+            system_clock::now())
+            .save();
+
+        auto points = std::vector<int>{5, 6, 7};
+        auto res = coll.update_one(MONGO_ODM_KEY(Bar::w) == 1010,
+                                   MONGO_ODM_KEY(Bar::arr).pull_all(points));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 1);
+
+        auto bar = Bar::find_one(MONGO_ODM_KEY(Bar::w) == 1010);
+        REQUIRE(bar);
+        REQUIRE((bar->arr == std::vector<int>{4}));
+    }
+
+    SECTION("Test $addToSet array update operator.", "[mongo_odm::Nvp::add_to_set]") {
+        Bar(1111, 10, 999, true, "addToSet", {1, 0}, {4, 5, 6}, {{9, 10}, {11, 12}},
+            system_clock::now())
+            .save();
+        // Add existing element to array, document is unchanged
+        auto res =
+            coll.update_one(MONGO_ODM_KEY(Bar::w) == 1111, MONGO_ODM_KEY(Bar::arr).add_to_set(4));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 0);
+
+        // Add new element to array, document is modified.
+        res = coll.update_one(MONGO_ODM_KEY(Bar::w) == 1111, MONGO_ODM_KEY(Bar::arr).add_to_set(7));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 1);
+
+        auto bar = Bar::find_one(MONGO_ODM_KEY(Bar::w) == 1111);
+        REQUIRE(bar);
+        REQUIRE((bar->arr == std::vector<int>{4, 5, 6, 7}));
+
+        // Add array of elements to array.
+        res = coll.update_one(MONGO_ODM_KEY(Bar::w) == 1111,
+                              MONGO_ODM_KEY(Bar::arr).add_to_set(std::vector<int>{7, 8, 9}));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 1);
+
+        bar = Bar::find_one(MONGO_ODM_KEY(Bar::w) == 1111);
+        REQUIRE(bar);
+        REQUIRE((bar->arr == std::vector<int>{4, 5, 6, 7, 8, 9}));
+    }
+
+    SECTION("Test $push array update operator.", "[mongo_odm::Nvp::push]") {
+        Bar(1212, 10, 999, true, "push", {1, 0}, {4, 5, 6}, {{9, 10}, {11, 12}},
+            system_clock::now())
+            .save();
+        auto res = coll.update_one(MONGO_ODM_KEY(Bar::w) == 1212, MONGO_ODM_KEY(Bar::arr).push(7));
+        REQUIRE(res);
+        REQUIRE(res->modified_count() == 1);
+
+        auto bar = Bar::find_one(MONGO_ODM_KEY(Bar::w) == 1212);
+        REQUIRE(bar);
+        REQUIRE((bar->arr == std::vector<int>{4, 5, 6, 7}));
+
+        // push several element at once
+        res = coll.update_one(MONGO_ODM_KEY(Bar::w) == 1212,
+                              MONGO_ODM_KEY(Bar::arr).push(std::vector<int>{7, 8, 9}));
```

Comment:
The compiler has trouble inferring the type of the initializer lists, since they can be any iterable.